### PR TITLE
Added additional formatting options for slices

### PR DESCRIPTION
**Pull request by Gilnaa**
First PR, sorry if anything is out of place.

Specifically I didn't have any idea about what to write in the "unstable" attribute, but I improvised and it seems to compile.

This PR tackles [1916](https://github.com/rust-lang/rfcs/issues/1916).
I thought about writing something similar to what builders::DebugInner does, but in the implementation for the rest of the traits, the Alternate flag is already taken (e.g. # in Debug is pretty-print, # in Octal is `0o`).

Cheers

### CODE REVIEW

**Comment by mattico** [comment]
Rustforge has some documentation about creating a new feature: https://forge.rust-lang.org/feature-guide.html

**Comment by Gilnaa** [comment]
Is this PR big enough to warrent a feature-flag? What is the usual procedure for small library changes?

**Comment by alexcrichton** [comment]
Thanks for the PR @Gilnaa! Unfortunately stability attributes right now don't cover trait implementations, so this would be "insta stable" in the sense of we *can't* mark this unstable to start out. In any case though this looks great! I'm personally a fan of these implementations.

Could you update the `#[unstable]` tag to `#[stable]` to reflect how they'll be stable on landing?

The next step here would then be to get a sign-off from the @rust-lang/libs team for approval, and to do that I'll...

@rfcbot fcp merge

**Comment by shepmaster** [comment]
You have a typo in your PR title (and maybe elsewhere?): `addtional`. I'll fix the PR title.

**Comment by Gilnaa** [comment]
@alexcrichton Fixed
@shepmaster Right, thanks :) . I couldn't find any more typos.

**Comment by bluss** [comment]
It's understood that this will use formatting options once per element? Just checking. For example padding, field width etc is per element, not for the whole slice.

For the record I think that's ok and it's the one issue that has been "unresolved" before adding this kind of adaptor for an iterator.

File: src/libcore/fmt/mod.rs [hunk 1]
```diff
@@ -1630,6 +1630,38 @@ impl<T: ?Sized> Debug for *mut T {
...
+        #[stable(feature = "additional_slice_formatting", since = "1.22.0")]
+        impl<T: $tr> $tr for [T] {
+            fn fmt(&self, f: &mut Formatter) -> Result {
+                let mut has_items = false;
+
+                write!(f, "[")?;
+
+                for o in self.iter() {
+                    if has_items {
+                        write!(f, ", ")?;
```
Comment (bluss) [inline src/libcore/fmt/mod.rs]: this can be f.write_str to avoid some dynamic dispatch.

Comment (bluss) [reply src/libcore/fmt/mod.rs]: (Same with the other single strings)

**Comment by Gilnaa** [comment]
@bluss Yes, the padding information is carried to each element using the Formater instance.

**Comment by bluss** [comment]
I think that's good. That kind of padding/width behaviour is indispensable in for example ndarray, to format a matrix of numbers in a readable way. The reminder was intended for everyone deciding on the issue, since that aspect hasn't been discussed yet (that I can see).

**Comment by ollie27** [comment]
What about `Display`? What about other collections?

This is committing to a completely arbitrary choice of surrounding with "[]" and separating with ", " which isn't what #33127 or #41674 wanted.

What I'd personally like to see is a nice way to format items in iterators in general like [`itertools::Itertools::format`](https://docs.rs/itertools/0.6.3/itertools/trait.Itertools.html#method.format) does.

**Comment by sfackler** [comment]
I kind of agree with @ollie27 - I would expect `{:x}` to just make a hex string for a byte slice if anything.

**Comment by Gilnaa** [comment]
@ollie27 I don't know about display. I have no idea what was the original reasoning *not* to impl it for slices, but I assume there was one.

@sfackler That would either require all [T] to be renderd like this, which is not ideal, or have a specialization for [u8], but that would cause inconsistency and violate the principle of least surprise.

Of course, I completely agree that there ought to do this hexdump-esque operation, but the domain of format specifiers is not that large. 
Maybe create somekind of wrapper newtype the implements it differently?

**Comment by dtolnay** [comment]
I agree with @ollie27 that this is a confusingly arbitrary choice of behavior to assign to `{:x}` and friends. I would prefer not to accept this, and solve it with adapter types in an external library instead.

```rust
format!("{:#x}", CommaSeparated(slice)) == "0x0, 0x1, 0x2"

format!("[{:#x}]", CommaSeparated(slice)) == "[0x0, 0x1, 0x2]"

format!("{:02x}", Concatenated(slice)) == "000102"
```

**Comment by Gilnaa** [comment]
I understand the motivation, but this just leave slice (directly) un-formattable using these traits?

**Comment by alexcrichton** [comment]
Ok we got a chance to discuss this during the libs triage meeting today, and our conclusion for now was that we probably don't actually want to merge it. As evidenced from the discussion it's not clear what the semantics should be here of these traits implemented directly on slices. @dtolnay's [suggestion](https://github.com/rust-lang/rust/pull/44751#issuecomment-331981315) as well could also possibly be turned into an entire library dedicated to providing high-quality formatting-related utilities for slices and other datatypes, handling information like formatting flags, customization, etc.

Our feeling was that we probably in the long run want a sort of "adapter" to implement this functionality in the standard library, one that you have to reach for and explicitly ask for so you know what it's doing. This we felt was best prototyped outside the standard library before moving in.

In light of all that I'm going to close this for now, but thanks regardless for the PR @Gilnaa!

**Pull request closed without merging**